read-fonts/src/tables/aat.rs

```diff
@@ -0,0 +1,619 @@
+//! Apple Advanced Typography common tables.
+//!
+//! See <https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6Tables.html>
+
+include!("../../generated/generated_aat.rs");
+
+/// Predefined classes.
+///
+/// See <https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6Tables.html>
+pub mod class {
+    pub const END_OF_TEXT: u8 = 0;
+    pub const OUT_OF_BOUNDS: u8 = 1;
+    pub const DELETED_GLYPH: u8 = 2;
+}
+
+/// Lookup segment for format 2.
+#[derive(Copy, Clone, bytemuck::AnyBitPattern)]
+#[repr(packed)]
+pub struct LookupSegment2<T>
+where
+    T: LookupValue,
+{
+    /// Last glyph index in this segment.
+    pub last_glyph: BigEndian<u16>,
+    /// First glyph index in this segment.
+    pub first_glyph: BigEndian<u16>,
+    /// The lookup value.
+    pub value: BigEndian<T>,
+}
+
+impl<T: LookupValue> FixedSize for LookupSegment2<T> {
+    const RAW_BYTE_LEN: usize = std::mem::size_of::<Self>();
+}
+
+impl<'a> Lookup2<'a> {
+    fn segments<T: LookupValue>(&self) -> Result<&[LookupSegment2<T>], ReadError> {
+        FontData::new(self.segments_data())
+            .cursor()
+            .read_array(self.n_units() as usize)
+    }
+}
+
+/// Lookup single record for format 6.
+#[derive(Copy, Clone, bytemuck::AnyBitPattern)]
+#[repr(packed)]
+pub struct LookupSingle<T>
+where
+    T: LookupValue,
+{
+    /// The glyph index.
+    pub glyph: BigEndian<u16>,
+    /// The lookup value.
+    pub value: BigEndian<T>,
+}
+
+impl<'a> Lookup6<'a> {
+    fn entries<T: LookupValue>(&self) -> Result<&[LookupSingle<T>], ReadError> {
+        FontData::new(self.entries_data())
+            .cursor()
+            .read_array(self.n_units() as usize)
+    }
+}
+
+impl<T: LookupValue> FixedSize for LookupSingle<T> {
+    const RAW_BYTE_LEN: usize = std::mem::size_of::<Self>();
+}
+
+pub struct TypedLookup<'a, T> {
+    lookup: Lookup<'a>,
+    _marker: std::marker::PhantomData<fn() -> T>,
+}
+
+impl<'a, T: LookupValue> TypedLookup<'a, T> {
+    /// Returns the value associated with the given index.
+    pub fn value(&self, index: u16) -> Result<T, ReadError> {
+        match &self.lookup {
+            Lookup::Format0(lookup) => {
+                let data = lookup.values_data();
```

Comment:
I'd be tempted to have a separate `value` method for each lookup type, and then this method would just forward calls? a bit more readable, and I often find that for some test or something I end up instantiating a specific format and want to be able to call methods on it..

Reply:
I waffled a bit on this because the type of the lookup is dependent on the containing table so the value methods on each variant would need a type parameter.

I don’t have strong feelings on this though and agree with your points so I’ll split them out.

Reply:
Done